### ISSUE TITLE
CA1416 Triggered even though platform validity check has been completed.

### ISSUE THREAD

**Issue by Zintom** [dotnet/runtime#77902]
### Description

I am working on a net6.0-android project. In this project I have a call-site which is reachable by all platform levels however there is some code which is only available on Android-O(26) and up.

I guard this code with the (recommend) check:
`if (Build.VERSION.SdkInt >= BuildVersionCodes.O)`

However the error [CA1416](https://learn.microsoft.com/en-gb/dotnet/fundamentals/code-analysis/quality-rules/ca1416) still shows in the IDE.

Compile time directive `#if ANDROID26_0_OR_GREATER` does not resolve this either.

### Reproduction Steps

Within a call-site reachable by all platforms, write a conditional statement which executes code for a specific platform or above.

![image](https://user-images.githubusercontent.com/18383281/200025936-cb4c168c-20c4-4ad0-be32-6b06e8d97a59.png)

### Expected behavior

No warning.

### Actual behavior

![image](https://user-images.githubusercontent.com/18383281/200025936-cb4c168c-20c4-4ad0-be32-6b06e8d97a59.png)

### Known Workarounds

None.

### Configuration

net6.0-android31.0
Minimum API 23
Target API 31

**Comment by Zintom** [comment]
Just tried on API-33, same issue.

**Comment by radical** [comment]
cc @mdh1418 @steveisok

**Comment by buyaa-n** [comment]
> I think this code patter is probably beynd the limits of the analyzer

Right, the analyzer only recognizes platform check operations provided by CoreLib.

You could move your custom platform check login into an API and annotate that API with `SupportedOSPlatformGuard("platform")` desired platform and optional version, then the analyzer could recognize that API as platform check operation.

```cs
[SupportedOSPlatformGuard("Android26.0")]
internal bool IsAndroid26 => Build.VERSION.SdkInt >= BuildVersionCodes.O;

// or could pass the version as a parameter
[SupportedOSPlatformGuard("Android")] // note that this option only added in .NET 7.0
internal bool IsAndroid(int version) => Build.VERSION.SdkInt >= version;
```
Then you could use in your code as a guard:
```cs
if (IsAndroid26)
{
    context.StartForegroundService(intent);
}
else
{
    context.StartService(intent);
}
// or
if (IsAndroid(26))
{
    context.StartForegroundService(intent);
}
else
{
    context.StartService(intent);
}
```

**Comment by Zintom** [comment]
@buyaa-n Thanks for these examples.

Wonderful, that is a lovely workaround.

**Comment by Zintom** [comment]
I've found myself in another predicament.

So the above code works for my first example.

However still throws in this scenario:
![image](https://user-images.githubusercontent.com/18383281/202220704-aa5eaea9-cab7-4af3-8935-fe65956ea626.png)

The analyser recognises that in the else scope of the code the API is less than 33 (`GetParcelableExtra(string)` is obseleted as of API 33 and without this guard throws its own warning), however it still warns in the first scope that the API is only accessible on API 33!

**Comment by akoeplinger** [comment]
@Zintom is there a reason why you can't use `if (OperatingSystem.IsAndroidVersionAtLeast(26)` instead?

**Comment by Zintom** [comment]
@akoeplinger I was not aware of this method and will try it.

**Comment by Zintom** [comment]
@akoeplinger I was not previously aware of `OperatingSystem.IsAndroidVersionAtLeast(int)`, this does have the functionality I am looking for.

My original issue is that the analyser should be able to realise when we are doing manual checks against `Build.VERSION.SdkInt`.

As @buyaa-n says, once the `SupportedOSPlatformGuard` attribute with the `int` version capability is working, this will solve the issue.
